### PR TITLE
Integrate with data transfer worker

### PR DESCRIPTION
**Pull request by RajaPiseD**
## Description
Changes to replace ansft calls with calls to data transfer service using python client of data transfer service

Data transfer binary is downloaded from data transfer server and started as a worker

Below operations are now performed using data transfer worker:

1. Upload of files (input files and execution scripts)
2. Download of files
3. Upload of project archive
4. Download of project archive

## Checklist
- [X] I have tested these changes locally.
- [X] I have added unit tests (if appropriate).
- [ ] I have added necessary documentation or updated existing documentation.
- [ ] I have linked the issue(s) addressed by this PR if any.

### CODE REVIEW

File: src/ansys/hps/client/client.py [hunk 1]
```diff
@@ -225,13 +236,39 @@ def __init__(
...
         warnings.warn(msg, DeprecationWarning)
         log.warning(msg)
         return self.url
 
+    def _start_dt_worker(self):
+
+        if self.dt_client is None:
+            log.info("Starting Data Transfer client.")
+            # start Data transfer client
+            self.dt_client = DTClient()
```
Comment (FedericoNegri) [inline src/ansys/hps/client/client.py:258]: You should set the `download_dir` to what we agreed for all clients:
- windows: ${LOCALAPPDATA}/Ansys/binaries/worker
- linux: ${HOME}/.ansys/binaries/worker
- macos: ask Michal what's preferable :-)

File: src/ansys/hps/client/client.py [hunk 2]
```diff
@@ -245,6 +282,8 @@ def _auto_refresh_token(self, response, *args, **kwargs):
             response.request.headers.update(
                 {"Authorization": self.session.headers["Authorization"]}
             )
+            if self.dt_client is not None:
+                self.dt_client.binary_config.update(token=self.access_token)
```
Comment (FedericoNegri) [inline src/ansys/hps/client/client.py:286]: This is not enough to make sure we capture token expiration issues when making requests to the worker api. We need to be able to inject a refresh hook function into the dt_client. Let's discuss it.

File: src/ansys/hps/client/jms/api/project_api.py [hunk 3]
```diff
@@ -657,17 +669,43 @@ def _delete_objects(self, objects: List[Object], obj_type: Type[Object]):
...
+        log.info(f"Operation {op[0].state}")
+        if op[0].state == OperationState.Succeeded:
+            for f in files:
+                if getattr(f, "hash", None) is not None:
+                    fpath = os.path.join(out_path, f"{f.id}")
+                    download_path = os.path.join(fpath, f.evaluation_path)
+                    with open(download_path, "rb") as inp:
+                        f.content = inp.read()
+        else:
+            log.error(f"Download of files failed")
```
Comment (FedericoNegri) [inline src/ansys/hps/client/jms/api/project_api.py:704]: let's raise here

File: src/ansys/hps/client/jms/api/project_api.py [hunk 4]
```diff
@@ -682,32 +720,54 @@ def get_files(project_api: ProjectApi, as_objects=True, content=False, **query_p
...
-        f.size = r.request.headers.get("Content-Length", None)
-
-        if is_file:
-            content.close()
+        op = project_api.client.dt_api.wait_for(op.id)
+        log.info(f"Operation {op[0].state}")
+        if op[0].state == OperationState.Succeeded:
+            _fetch_file_metadata(project_api, files, dsts)
+        else:
+            log.error(f"Upload of files failed")
```
Comment (FedericoNegri) [inline src/ansys/hps/client/jms/api/project_api.py:749]: let's raise here

File: src/ansys/hps/client/jms/api/project_api.py [hunk 5]
```diff
@@ -755,21 +815,26 @@ def _download_file(
...
+    log.info(f"Downloading file {file.id}")
+    src = StoragePath(path=f"{base_dir}/{os.path.basename(file.storage_id)}")
+    dst = StoragePath(path=download_path, remote="local")
+    op = project_api.client.dt_api.copy([SrcDst(src=src, dst=dst)])
+    op = project_api.client.dt_api.wait_for([op.id])
+
+    log.info(f"Operation {op[0].state}")
+
+    if op[0].state != OperationState.Succeeded:
+        log.error(f"Download of file {file.evaluation_path} with id {file.id} failed")
```
Comment (FedericoNegri) [inline src/ansys/hps/client/jms/api/project_api.py:836]: let's raise

File: src/ansys/hps/client/jms/api/project_api.py [hunk 6]
```diff
@@ -755,21 +815,26 @@ def _download_file(
     stream: bool = True,
```
Comment (FedericoNegri) [inline src/ansys/hps/client/jms/api/project_api.py:815]: - `stream` is unused now
- progress tracking need to be implemented using the operation progress field, will be a breaking change.

File: src/ansys/hps/client/client.py (hunk 1 above)
Comment (RajaPiseD) [reply src/ansys/hps/client/client.py:258]: updated code to use proper directory for downloading binary

File: src/ansys/hps/client/jms/api/project_api.py (hunk 3 above)
Comment (RajaPiseD) [reply src/ansys/hps/client/jms/api/project_api.py:704]: updated code

File: src/ansys/hps/client/jms/api/project_api.py (hunk 4 above)
Comment (RajaPiseD) [reply src/ansys/hps/client/jms/api/project_api.py:749]: updated code

File: src/ansys/hps/client/client.py [hunk 7]
```diff
@@ -128,6 +136,7 @@ def __init__(
         all_fields=True,
         verify: Union[bool, str] = None,
         disable_security_warnings: bool = True,
+        data_transfer_url: str | None = None,
```
[the remark this answers is not in the document]
Comment (RajaPiseD) [reply src/ansys/hps/client/client.py:139]: updated code to use url to derive dts url

File: src/ansys/hps/client/client.py [hunk 8]
```diff
@@ -225,13 +236,39 @@ def __init__(
         self._unauthorized_num_retry = 0
         self._unauthorized_max_retry = 1
 
+        def exit_handler():
+            log.info("Exiting gracefully.")
```
[the remark this answers is not in the document]
Comment (RajaPiseD) [reply src/ansys/hps/client/client.py:240]: updated code

File: src/ansys/hps/client/jms/api/project_api.py (hunk 5 above)
Comment (RajaPiseD) [reply src/ansys/hps/client/jms/api/project_api.py:836]: updated code

File: .github/workflows/ci_cd.yml [hunk 9]
```diff
@@ -120,6 +126,12 @@ jobs:
...
 
     steps:
+      - uses: extractions/netrc@v2
+        with:
+          machine: github.com
+          username: pyansys-ci-bot
+          password: ${{ secrets.PYANSYS_CI_BOT_PACKAGE_TOKEN }}
+
       - name: Build wheelhouse and perform smoke test
         uses: ansys/actions/build-wheelhouse@v7
```
Comment (RobPasMue) [inline .github/workflows/ci_cd.yml:136]: This action allows you to skip the license check if you want to - but please add a TODO to solve this eventually: https://actions.docs.ansys.com/version/stable/build-actions/index.html#build-wheelhouse-action

File: src/ansys/hps/client/__version__.py [hunk 10]
```diff
@@ -29,6 +29,10 @@
...
 __version__ = importlib_metadata.version("ansys-hps-client")
 
+__company__ = "Ansys Switzerland GmbH"
+
+__company_short__ = "Ansys"
+
 # this is only a convenience to default the version
 # of Ansys simulation applications in PyHPS examples
-__ansys_apps_version__ = "2024 R1"
+__ansys_apps_version__ = "2024 R2"
```
Comment (FedericoNegri) [inline src/ansys/hps/client/__version__.py:38]: bumping apps version is release related, should not be done in this PR

File: src/ansys/hps/client/__version__.py [hunk 11]
```diff
@@ -29,6 +29,10 @@
 # major, minor, patch
 __version__ = importlib_metadata.version("ansys-hps-client")
 
+__company__ = "Ansys Switzerland GmbH"
+
+__company_short__ = "Ansys"
+
```
Comment (FedericoNegri) [inline src/ansys/hps/client/__version__.py:35]: Please remove these, this is a pyansys package not an internal one.

File: src/ansys/hps/client/client.py [hunk 12]
```diff
@@ -98,15 +105,17 @@ class Client(object):
     >>> cl = Client(
     ...     url="https://localhost:8443/hps",
     ...     username="repuser",
-    ...     password="repuser"
+    ...     password="repuser",
+            dts_url="https://localhost:8443/hps/dt/api/v1"
```
Comment (FedericoNegri) [inline src/ansys/hps/client/client.py:109]: this need to be removed

File: src/ansys/hps/client/client.py [hunk 13]
```diff
@@ -98,15 +105,17 @@ class Client(object):
...
     ... )
 
     Create a client object and connect to HPS with a refresh token.
 
     >>> cl = Client(
     ...     url="https://localhost:8443/hps",
     ...     username="repuser",
-    ...     refresh_token="eyJhbGciOiJIUzI1NiIsInR5cC..."
+    ...     refresh_token="eyJhbGciOiJIUzI1NiIsInR5cC...",
+            dts_url="https://localhost:8443/hps/dt/api/v1"
```
Comment (FedericoNegri) [inline src/ansys/hps/client/client.py:118]: this need to be removed

File: src/ansys/hps/client/client.py [hunk 14]
```diff
@@ -280,6 +298,62 @@ def rep_url(self) -> str:
...
+                )
+                self.dt_client.start()
+
+                self.dt_api = DataTransferApi(self.dt_client)
+                self.dt_api.status(wait=True)
+            except Exception as ex:
+                log.debug(ex)
+                raise HPSError("Error occurred when starting Data Transfer client.")
+
+    def _get_download_dir(self, company=None):
```
Comment (FedericoNegri) [inline src/ansys/hps/client/client.py:324]: no need to parametrize the company name, just use ansys/Ansys

File: src/ansys/hps/client/jms/api/jms_api.py [hunk 15]
```diff
@@ -74,12 +76,12 @@ def url(self) -> str:
...
-        """URL of the file storage gateway."""
-        if self._fs_url is None:
-            self._fs_url = _find_available_fs_url(self.get_storage())
-        return self._fs_url
+    # @property
+    # def fs_url(self) -> str:
+    #     """URL of the file storage gateway."""
+    #     if self._fs_url is None:
+    #         self._fs_url = _find_available_fs_url(self.get_storage())
+    #     return self._fs_url
```
Comment (FedericoNegri) [inline src/ansys/hps/client/jms/api/jms_api.py:84]: please remove commented code

File: src/ansys/hps/client/jms/api/jms_api.py [hunk 16]
```diff
@@ -449,44 +444,65 @@ def _restore_project(jms_api, archive_path):
...
+#         except Exception as ex:
+#             log.debug(ex)
+#             continue
+#         if r.status_code == 200 and is_ansft:
+#             return url
+#
+#     raise HPSError(
+#         f"All defined file storage gateways are unavailable"
+#         f" ({', '.join([d['url'] for d in rest_gateways])})."
+#     )
```
Comment (FedericoNegri) [inline src/ansys/hps/client/jms/api/jms_api.py:508]: please remove commented code

File: src/ansys/hps/client/jms/api/project_api.py [hunk 17]
```diff
@@ -657,17 +674,44 @@ def _delete_objects(self, objects: List[Object], obj_type: Type[Object]):
...
     """
-    Download files directly using the fs REST gateway.
+    Download files directly using data transfer worker.
 
-    This is a temporary implementation for downloading files. It is to be
-    replaced with direct ansft calls, when it is available as a Python package.
     """
 
+    project_api.client._start_dt_worker()
+    out_path = os.path.join(os.path.dirname(__file__), "downloads")
```
Comment (FedericoNegri) [inline src/ansys/hps/client/jms/api/project_api.py:682]: Can we use a proper temp dir instead? https://docs.python.org/3/library/tempfile.html#tempfile.TemporaryDirectory

File: src/ansys/hps/client/jms/api/project_api.py [hunk 18]
```diff
@@ -682,32 +726,69 @@ def get_files(project_api: ProjectApi, as_objects=True, content=False, **query_p
...
-            f"{project_api.fs_bucket_url}/{f.storage_id}",
-            data=content,
-            headers=fs_headers,
+        filePath = f.src
+        if isinstance(f.src, io.IOBase):
+            tempFiles.append(f.storage_id)
+            mode = "wb" if isinstance(f.src, io.BytesIO) else "w"
+            with open(f.storage_id, mode) as out:
+                out.write(f.src.getvalue())
+                filePath = f.storage_id
```
Comment (FedericoNegri) [inline src/ansys/hps/client/jms/api/project_api.py:749]: please write these files into a temp directory

File: .github/actions/hps_services/action.yml [hunk 19]
```diff
@@ -43,15 +43,25 @@ runs:
...
+#    - if: ${{ inputs.version == 'latest-dev' }}
+#      uses: dsaltares/fetch-gh-release-asset@master
+#      name: Download Internal Services Artifact
+#      with:
+#        file: 'docker-compose-internal.tar.gz'
+#        target: 'docker-compose-artifact/docker-compose-internal.tar.gz'
+#        repo: ansys-internal/rep-deployments@mpawlik/dt_integration
+#        version: tags/latest-dev
+#        token: ${{ inputs.token }}
+    - uses: dawidd6/action-download-artifact@v3
```
Comment (FedericoNegri) [inline .github/actions/hps_services/action.yml:55]: I think we could in general use this pattern:
- for released versions we get the artifacts from the release assets
- for latest-dev, we get artifacts from the latest successful build

To support testing against features, we should expose a new input `feature` (only used when `version=latest-dev`) to be passed as  branch to dawidd6/action-download-artifact.

File: .github/actions/hps_services/action.yml [hunk 20]
```diff
@@ -80,17 +90,19 @@ runs:
...
       run: |
+        echo "$(pwd)"
+        pwd
         ls -la
         tar -xvzf docker-compose-internal.tar.gz
         cd docker-compose
-        FSGATEWAY_TAG=latest-dev JMS_TAG=latest-dev docker-compose pull
-        FSGATEWAY_TAG=latest-dev JMS_TAG=latest-dev docker-compose up -d
+        FSGATEWAY_TAG=latest-dev JMS_TAG=mgunupa-integration DTS_TAG=latest-dev docker-compose pull
+        FSGATEWAY_TAG=latest-dev JMS_TAG=mgunupa-integration DTS_TAG=latest-dev docker-compose --profile=stable --profile=data-transfer up -d
```
Comment (FedericoNegri) [inline .github/actions/hps_services/action.yml:99]: I think by now we should be able to remove the tags as well the profiles

File: src/ansys/hps/client/client.py (hunk 2 above)
Comment (FedericoNegri) [reply src/ansys/hps/client/client.py:384]: Agreed to follow up on this at a later time https://github.com/ansys/pyhps/issues/452

File: src/ansys/hps/client/jms/api/jms_api.py (hunk 16 above)
Comment (RajaPiseD) [reply src/ansys/hps/client/jms/api/jms_api.py:508]: updated code

File: src/ansys/hps/client/__version__.py (hunk 11 above)
Comment (RajaPiseD) [reply src/ansys/hps/client/__version__.py:35]: reverted changes

File: src/ansys/hps/client/client.py (hunk 13 above)
Comment (RajaPiseD) [reply src/ansys/hps/client/client.py:118]: line removed

File: src/ansys/hps/client/__version__.py (hunk 10 above)
Comment (RajaPiseD) [reply src/ansys/hps/client/__version__.py:38]: reverted the change

File: src/ansys/hps/client/jms/api/project_api.py (hunk 18 above)
Comment (RajaPiseD) [reply src/ansys/hps/client/jms/api/project_api.py:749]: updated code

File: src/ansys/hps/client/jms/api/project_api.py (hunk 17 above)
Comment (RajaPiseD) [reply src/ansys/hps/client/jms/api/project_api.py:682]: updated code

File: src/ansys/hps/client/client.py (hunk 14 above)
Comment (RajaPiseD) [reply src/ansys/hps/client/client.py:321]: updated code

**Review by FedericoNegri — APPROVED** [review]

File: pyproject.toml [hunk 21]
```diff
@@ -31,7 +31,8 @@ dependencies = [
     "marshmallow_oneofschema>=2.0.1",
     "backoff>=2.0.0",
     "pydantic>=1.10.0",
-    "PyJWT>=2.8.0"
+    "PyJWT>=2.8.0",
+    "ansys-hps-data-transfer-client@git+https://github.com/ansys-internal/hps-data-transfer-client.git@main#egg=ansys-hps-data-transfer-client"
```
Comment (jorgepiloto) [inline pyproject.toml:35]: @FedericoNegri @RajaPiseD I want to bring to your attention that adding dependencies using the git specifier is not valid when uploading packages to PyPI. The Python Package Index will reject any uploads containing dependencies specified like this.

Comment (FedericoNegri) [reply pyproject.toml:35]: @jorgepiloto we're aware, that's currently an internal dependency which is meant to either be incorporated in pyhps directly or become an open source pyansys package on its own. Either way, it's a prerequisite for the next pyhps release.

**Pull request merged**